### PR TITLE
Adding Boruvka algorithm + tests for Minimum Spanning Trees

### PR DESCRIPTION
**Pull request by FranciscoThiesen**
#### Checklist

- [x] Added description of change
- [x] Added file name matches [File name guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#New-File-Name-guidelines)
- [x] Added tests and example, test must pass
- [x] Added documentation so that the program is self-explanatory and educational - [Doxygen guidelines](https://www.doxygen.nl/manual/docblocks.html)
- [x] Relevant documentation/comments is changed or added
- [x] PR title follows semantic [commit guidelines](https://github.com/TheAlgorithms/C-Plus-Plus/blob/master/CONTRIBUTING.md#Commit-Guidelines)
- [x] Search previous suggestions before making a new one, as yours may be a duplicate.
- [x] I acknowledge that all my contributions will be made under the project's license.

Notes:

### CODE REVIEW

**Comment by FranciscoThiesen** [comment]
@Panquesito7 @realstealthninja reviews appreciated, thank you!

File: graph/boruvka.cpp [hunk 1]
```diff
@@ -0,0 +1,191 @@
...
+};
+
+/**
+ * @brief Boruvka's algorithm to compute the minimum spanning tree cost.
+ * @param num_nodes Number of nodes in the graph.
+ * @param num_edges Number of edges in the graph.
+ * @param edges Edges of the graph.
+ * @return Minimum spanning tree cost.
+ */
+long long boruvka(
```
Comment (realstealthninja) [inline graph/boruvka.cpp:64]:
```suggestion
int64_t boruvka(
```

File: graph/boruvka.cpp [hunk 2]
```diff
@@ -0,0 +1,191 @@
...
+    int num_nodes, int num_edges,
```
Comment (realstealthninja) [inline graph/boruvka.cpp:65]: wouldnt it make more sense to make these unsigned since num_nodes/num_edges would never be negative?

File: graph/boruvka.cpp [hunk 3]
```diff
@@ -0,0 +1,191 @@
...
+
+/**
+ * @brief Test function for Boruvka's algorithm.
+ */
+void tests() {
+    std::vector<std::pair<int, std::pair<int, int>>> edges;
+    edges.push_back({1, {0, 1}});
+    edges.push_back({2, {0, 2}});
+    edges.push_back({3, {1, 2}});
+    edges.push_back({4, {1, 3}});
```
Comment (realstealthninja) [inline graph/boruvka.cpp:134]: as mentioned above if you really want to use int then might want to add tests for cases when some of the values are negative

**Review by realstealthninja — CHANGES_REQUESTED** [review]

File: graph/boruvka.cpp (hunk 2 above)
Comment (FranciscoThiesen) [reply graph/boruvka.cpp:65]: Will change it. Thanks for pointing this out.

Comment (FranciscoThiesen) [reply graph/boruvka.cpp]: You are absolutely right! Just changed it.

**Comment by FranciscoThiesen** [comment]
@realstealthninja I've addressed your suggestions. Also changed from int -> uint32_t in the union find, where negative values also don't make sense.

Let me know if there are any other issues.

File: graph/boruvka.cpp [hunk 4]
```diff
@@ -0,0 +1,191 @@
...
+    edges.push_back({5, {2, 3}});
+    edges.push_back({6, {2, 4}});
+    edges.push_back({7, {3, 4}});
+    assert(boruvka(5, 7, edges) == 9);
+}
+
+int main() {
+    tests();
```
Comment (realstealthninja) [inline graph/boruvka.cpp:142]:
```suggestion
    tests(); // run self-test implementations
```

File: graph/boruvka.cpp [hunk 5]
```diff
@@ -0,0 +1,191 @@
...
+ * tree of a graph.
+ * @author FranciscoThiesen
+ * @details For more details, see
+ * https://en.wikipedia.org/wiki/Bor%C5%AFvka%27s_algorithm
+ */
+
+#include <cassert>
+#include <iostream>
+#include <numeric>
+#include <vector>
```
Comment (realstealthninja) [inline graph/boruvka.cpp:13]: The purpose of these headers must be documented
A simple comment would suffice.

```cpp
#include <header> /// for <reason>
```

**Review by realstealthninja — CHANGES_REQUESTED** [review]
Thank you for your contributions @Panquesito7 could you have a look as well?

File: graph/boruvka.cpp (hunk 5 above)
Comment (FranciscoThiesen) [reply graph/boruvka.cpp]: Done!

**Comment by FranciscoThiesen** [comment]
@realstealthninja iterated on the last round of reviews!

**Comment by FranciscoThiesen** [comment]
@realstealthninja can you approve the workflows?

Thank you,
Francisco

**Comment by FranciscoThiesen** [comment]
@realstealthninja @Panquesito7 can you please approve the required workflows? I've addressed all the comments/suggested changes.

Thank you,
Francisco

**Comment by FranciscoThiesen** [comment]
@realstealthninja @Panquesito7

**Comment by FranciscoThiesen** [comment]
@realstealthninja @Panquesito7 any updates? I'd like to merge this if possible

**Comment by FranciscoThiesen** [comment]
@realstealthninja @Panquesito7

**Comment by FranciscoThiesen** [comment]
@realstealthninja @Panquesito7 can you kindly take a look? I have addressed all the changes requested and this is ready to merge. I can't initiate the checks and I am not currently a contributor.
